src/VASSAL/counters/BasicPiece.java

```diff
@@ -571,6 +612,13 @@ public String getState() {
     final Point p = getPosition();
     se.append(p.x).append(p.y);
     se.append(getGpId());
+    se.append(persistentProps.size());
+    // Persistent Property values will always be String (for now).
+    for (Object key : persistentProps.keySet()) {
+      se.append(key.toString()); 
```

Comment:
will throw an NPE if key is null, which is a valid key in HashMap

Reply:
Yep, plus the code was wrong, It should have been checking val for null.